File: microsynth/microsynth/report/open_label_orders/open_label_orders.js
// Copyright (c) 2022-2024, Microsynth, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Open Label Orders"] = {
    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "reqd": 1,
            "default": frappe.defaults.get_user_default("company") || frappe.defaults.get_global_default("company")
        }
    ],
    "onload": (report) => {
        hide_chart_buttons();
        report.page.add_inner_button( __("Pick labels"), function() {
            queue_builder(report.get_values());
        });
        report.page.add_inner_button( __("PrioPick"), function() {
            prio_pick(report.get_values());
        });
    }
};

function queue_builder(filters) {
    frappe.call({
        'method': "microsynth.microsynth.report.open_label_orders.open_label_orders.picking_ready",
        'callback': function(r) {
            if (r.message) {
                frappe.call({
                    'method': "microsynth.microsynth.report.open_label_orders.open_label_orders.get_data",
                    'args': {
                        'filters': {
                            'company': filters.company
                        }
                    },
                    'callback': function(r) {
                        // reset queue
                        locals.label_queue = r.message;
                        // add status flags to each entry
                        for (var i = 0; i < locals.label_queue.length; i++) {
                            locals.label_queue[i].status = 0;
                        }
                        // start queue processing
                        process_queue()
                    }
                });
            } else {
                warn_process_running();
            }
        }
    });
}

function warn_process_running() {
    frappe.msgprint( __("Another picking process is still in progress, please wait."), __("Picking in progress") );
}

function process_queue() {
    if (locals.label_queue.length > 0) {
        // process first entry
        label_order = locals.label_queue[0];
        if (label_order.additional_items) {
            let allowed_additional_items = ['20050'];
            for (let i = 0; i < label_order.additional_item_codes.length; i++) {
                let item_code = label_order.additional_item_codes[i];
                if (!allowed_additional_items.includes(item_code)) {
                    frappe.msgprint({
                        title: __("Unallowed additional Item"),
                        indicator: 'red',
                        message: __("Label Order " + label_order.sales_order + " contains the unallowed additional Item ") + item_code + ".<br>Label picking for this order is <b>not</b> permitted.<br>Please contact its creator " + label_order.owner + "."
                    });
                    return;
                } else {
                    frappe.msgprint({
                        title: __("Additional Items to consider"),
                        indicator: 'blue',
                        message: __("Label Order " + label_order.sales_order + " contains the following additional Items:<br><br>") + label_order.additional_items
                    });
                }
            }
        }
        if (label_order.status === 0) {
            // nothing done - first barcode
            first_barcode_dialog();
        } else if (label_order.status === 1) {
            // first barcode done - get second barcode
            second_barcode_dialog();
        } else {
            // create delivery note / pdf
                frappe.call({
                    'method': "microsynth.microsynth.report.open_label_orders.open_label_orders.pick_labels",
                    'args': {
                        'sales_order': locals.label_queue[0].sales_order,
                        'from_barcode': Number(locals.label_queue[0].from_barcode),
                        'to_barcode': locals.label_queue[0].to_barcode,
                        'number_length': locals.label_queue[0].from_barcode.length
                    },
                    'callback': function(r) {
                        // open print dialog & print
                        console.log(r.message);
                        //if (r.message.includes("DN-")){
                        window.open("/printview?doctype=Delivery%20Note&name=" + r.message + "&trigger_print=1&format=Delivery%20Note%20Sequencing%20Labels&no_letterhead=0&_lang=en", '_blank').focus();
                        //} else {
                        //    frappe.show_alert(r.message);
                        //}
                    }
                });
            // kick first order out and resume
            locals.label_queue.shift();
            process_queue();
        }
    } else {
        // queue is empty, all done ;-) refresh view to prevent duplicate pick
        frappe.query_report.refresh();
    }
}


function first_barcode_dialog() {
    if (locals.label_queue.length > 0) {
        // show scan first dialog
        frappe.prompt(
            [
                {
                    'fieldname': 'from_barcode',
                    'fieldtype': 'Data',
                    'label': 'From barcode'
                },
                {
                    'fieldname': 'sales_order',
                    'fieldtype': 'Link',
                    'label': 'Sales Order',
                    'read_only': 1,
                    'options': 'Sales Order',
                    'default': locals.label_queue[0].sales_order
                },
                {
                    'fieldname': 'item',
                    'fieldtype': 'Data',
                    'label': __('Item'),
                    'read_only': 1,
                    'default': locals.label_queue[0].qty + "x " + locals.label_queue[0].item_code + ": " + locals.label_queue[0].item_name
                },
                {
                    'fieldname': 'prefix',
                    'fieldtype': 'Data',
                    'label': 'Prefix',
                    'read_only': 1,
                    'default': locals.label_queue[0].prefix
                },
                {
                    'fieldname': 'range',
                    'fieldtype': 'Data',
                    'label': 'Range',
                    'read_only': 1,
                    'default': locals.label_queue[0].range
                }
            ],
            function (values) {
                // validation: if it fails, leave status on 0 and continue, on success move to 1
                var validated = false;
                // delete spaces and numbers to get the prefix
                var prefix = values.from_barcode.replace(/\s+/g, '').replace(/[0-9]/g, '');
                // delete spaces and letters
                var from_barcode = values.from_barcode.replace(/\s+/g, '').replace(/[^0-9]/g, '');
                var to_barcode = Number(from_barcode) + Number(locals.label_queue[0].qty) - 1;

                validated = is_in_range(locals.label_queue[0].range, Number(from_barcode))
                            && (!locals.label_queue[0].prefix || prefix == locals.label_queue[0].prefix);

                if (validated) {
                    locals.label_queue[0].status = 1;
                }
                else {
                    frappe.msgprint("invalid barcode range or invalid prefix");
                }
                locals.label_queue[0].from_barcode = from_barcode;

                //process_queue(); // check availability, the proceed
                are_labels_available(locals.label_queue[0].item_code, Number(from_barcode), to_barcode);
            },
            __("Pick first label"),
            __("OK")
        )
    }
}

function second_barcode_dialog() {
    if (locals.label_queue.length > 0) {
        // show scan second dialog
        frappe.prompt(
            [
                {
                    'fieldname': 'to_barcode',
                    'fieldtype': 'Data',
                    'label': 'To barcode'
                },
                {
                    'fieldname': 'print_shipping_label',
                    'fieldtype': 'Check',
                    'label': 'Print Shipping Label',
                    'default': (['3100', '3120', '3200', '3236', '3240', '3251'].includes(locals.label_queue[0].item_code) || locals.label_queue[0].additional_item_codes.includes('20050')) ? 1 : 0
                },
                {
                    'fieldname': 'sales_order',
                    'fieldtype': 'Link',
                    'label': 'Sales Order',
                    'read_only': 1,
                    'options': 'Sales Order',
                    'default': locals.label_queue[0].sales_order
                },
                {
                    'fieldname': 'item',
                    'fieldtype': 'Data',
                    'label': __('Item'),
                    'read_only': 1,
                    'default': locals.label_queue[0].qty + "x " + locals.label_queue[0].item_code + ": " + locals.label_queue[0].item_name
                },
                {
                    'fieldname': 'prefix',
                    'fieldtype': 'Data',
                    'label': 'Prefix',
                    'read_only': 1,
                    'default': locals.label_queue[0].prefix
                },
                {
                    'fieldname': 'range',
                    'fieldtype': 'Data',
                    'label': 'Range',
                    'read_only': 1,
                    'default': locals.label_queue[0].range
                },
                {
                    'fieldname': 'from_barcode',
                    'fieldtype': 'Data',
                    'label': 'From barcode',
                    'read_only': 1,
                    'default': (locals.label_queue[0].prefix ? locals.label_queue[0].prefix : '') + locals.label_queue[0].from_barcode
                }
            ],
            function (values) {
                // validation: if it fails, leave status on 0 and continue, on success move to 1
                var validated = false;
                // delete spaces and numbers to get prefix
                var prefix = values.to_barcode.replace(/\s+/g, '').replace(/[0-9]/g, '');
                // delete spaces and letters to get to_barcode as number
                var to_barcode = Number(values.to_barcode.replace(/\s+/g, '').replace(/[^0-9]/g, ''));

                validated =
                    is_in_range(locals.label_queue[0].range, to_barcode)
                    && to_barcode == Number(locals.label_queue[0].from_barcode) + Number(locals.label_queue[0].qty) - 1
                    && (!locals.label_queue[0].prefix || prefix == locals.label_queue[0].prefix);

                if (validated) {
                    frappe.show_alert("Barcodes OK");
                    locals.label_queue[0].status = 2;
                }
                else {
                    frappe.msgprint("invalid barcode range or invalid prefix");
                }
                locals.label_queue[0].to_barcode = to_barcode;

                if (values.print_shipping_label) {
                    // print shipping label
                    frappe.call({
                        "method": "microsynth.microsynth.labels.print_shipping_label",
                        "args": {
                            "sales_order_id": locals.label_queue[0].sales_order
                        }
                    });
                }

                process_queue();
            },
            __("Pick last label"),
            __("OK")
        )
    }
}

function is_in_range(ranges, value) {
    var range_list = ranges.split(',');

    for (var i = 0; i < range_list.length; i++) {
        var e = range_list[i].split('-')
        var start = e[0].trim()
        var end = e[1].trim()

        if ( start <= value && value <= end ){
            return true;
        }
    }
    return false;
}

function are_labels_available(item_code, from_barcode, to_barcode) {
    frappe.call({
        'method': "microsynth.microsynth.report.open_label_orders.open_label_orders.are_labels_available",
        'args': {
            'item_code': item_code,
            'from_barcode': Number(from_barcode),
            'to_barcode': to_barcode
        },
        'asyc': false,
        'callback': function(r) {
            if (r.message == 1) {
                // labels available, proceed
                process_queue()
            } else {
                // labels have already been used -> stop
                frappe.msgprint("At least one of the labels from " + Number(from_barcode) + " to " + to_barcode + " is already used.");
            }
        }
    });
}

function prio_pick(report_data) {
    frappe.call({
        'method': "microsynth.microsynth.report.open_label_orders.open_label_orders.picking_ready",
        'callback': function(r) {
            if (r.message) {
                // collect all sales orders
                var orders = [];
                for (var i = 0; i < frappe.query_report.data.length; i++) {
                    orders.push(frappe.query_report.data[i].sales_order);
                }
                // show order selection dialog
                frappe.prompt([
                        {
                            'fieldname': 'sales_order',
                            'fieldtype': 'Select',
                            'label': 'Sales Order',
                            'reqd': 1,
                            'options': orders.join("\n"),
                            'default': orders[0]
                        }
                    ],
                    function(values){
                        // reset and prepare queue
                        locals.label_queue = [];
                        // find order details
                        order_details = null;
                        for (var i = 0; i < frappe.query_report.data.length; i++) {
                            if (frappe.query_report.data[i].sales_order === values.sales_order) {
                                locals.label_queue.push(frappe.query_report.data[i]);
                                locals.label_queue[0].status = 0;
                                break;
                            }
                        }
                        // start queue
                        process_queue();
                    },
                    'Select Sales Order',
                    'OK'
                );
            } else {
                warn_process_running();
            }
        }
    });
}

File: microsynth/microsynth/report/open_label_orders/open_label_orders.py
# Copyright (c) 2022, Microsynth, libracore and contributors and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import os
import frappe
from frappe import _
import json
from erpnext.selling.doctype.sales_order.sales_order import make_delivery_note
from microsynth.microsynth.naming_series import get_naming_series
from datetime import datetime

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {"label": _("Sales Order"), "fieldname": "sales_order", "fieldtype": "Link", "options": "Sales Order", "width": 120},
        {"label": _("Web ID"), "fieldname": "web_order_id", "fieldtype": "Data", "width": 70},
        {"label": _("Punchout"), "fieldname": "is_punchout", "fieldtype": "Check", "width": 75},
        {"label": _("Customer"), "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 70},
        {"label": _("Customer name"), "fieldname": "customer_name", "fieldtype": "Data", "width": 150},
        {"label": _("Contact"), "fieldname": "contact", "fieldtype": "Link", "options": "Contact", "width": 60},
        {"label": _("Contact name"), "fieldname": "contact_name", "fieldtype": "Data", "width": 150},
        {"label": _("Date"), "fieldname": "date", "fieldtype": "Date", "width": 80},
        {"label": _("Item"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 200},
        {"label": _("Qty"), "fieldname": "qty", "fieldtype": "Integer", "width": 50},
        # {"label": _("Range"), "fieldname": "range", "fieldtype": "Data", "width": 80},
        {"label": _("Comment"), "fieldname": "comment", "fieldtype": "Data", "width": 100}
    ]

@frappe.whitelist()
def get_data(filters=None):
    if type(filters) == str:
        filters = json.loads(filters)
    elif type(filters) == dict:
        pass
    else:
        filters = dict(filters)

    open_label_orders = frappe.db.sql("""
        SELECT 
            `tabSales Order`.`name` AS `sales_order`,
            `tabSales Order`.`web_order_id` AS `web_order_id`,
            `tabSales Order`.`is_punchout` AS `is_punchout`,
            `tabSales Order`.`customer` AS `customer`,
            `tabSales Order`.`customer_name` AS `customer_name`,
            `tabSales Order`.`contact_person` AS `contact`,
            `tabSales Order`.`contact_display` AS `contact_name`,
            `tabSales Order`.`transaction_date` AS `date`,
            `tabSales Order Item`.`item_code` AS `item_code`,
            `tabSales Order Item`.`item_name` AS `item_name`,
            `tabSales Order Item`.`qty` AS `qty`,
            `tabLabel Range`.`range` AS `range`,
            `tabSales Order`.`comment` AS `comment`
        FROM `tabSales Order`
        LEFT JOIN `tabSales Order Item` ON
            (`tabSales Order Item`.`parent` = `tabSales Order`.`name` AND `tabSales Order Item`.`idx` = 1)
        LEFT JOIN `tabLabel Range` ON
            (`tabLabel Range`.`item_code` = `tabSales Order Item`.`item_code`)
        WHERE 
            `tabSales Order`.`product_type` = "Labels"
            AND `tabSales Order`.`docstatus` = 1
            AND `tabSales Order`.`company` = "{company}"
            AND `tabSales Order`.`transaction_date` > '2022-12-22'
            AND `tabSales Order`.`hold_order` <> 1
            AND NOT EXISTS (SELECT *
                            FROM `tabDelivery Note Item`
                            WHERE `tabDelivery Note Item`.`against_sales_order` = `tabSales Order`.`name`
                            AND `tabDelivery Note Item`.`docstatus` <> 2)
        ORDER BY `tabSales Order Item`.`item_code`, `tabSales Order`.`transaction_date` ASC;
    """.format(company=filters.get("company")), as_dict=True)
    
    return open_label_orders

ASSIGNMENT_HEADER = """Person_ID\tSales_Order\tWeb_Order_ID\tItem\tStart\tEnd\n"""
ASSIGNMENT_FIELDS = """{person_id}\t{sales_order}\t{web_order_id}\t{item}\t{start}\t{end}\n"""

def write_assignment_file(data):
    path = frappe.get_value("Sequencing Settings", "Sequencing Settings", "label_export_path")
    assignment_file = "{path}/{sales_order}.tab".format(path = path, sales_order=data["sales_order"])
    if os.path.exists(assignment_file):
        message = "<b>Sequencing label assignment file already exists:</b><br>" + assignment_file
        frappe.log_error(message, "Open label orders.write_assignment_file")
        frappe.throw(message)
    else:
        file = open(assignment_file, "w")
        file.write(ASSIGNMENT_HEADER)
        line = ASSIGNMENT_FIELDS.format(
            person_id = data["person_id"],
            sales_order = data["sales_order"],
            web_order_id = data["web_order_id"],
            item = data["item"],
            start = data["from_barcode"],
            end = data["to_barcode"]
        )
        file.write(line)
        file.close()
    return


# @frappe.whitelist()
# def pick_labels_without_timeout(sales_order, from_barcode, to_barcode):
#     """
#     Wrapper to call function pick_labels with a timeout > 120 seconds (here 600 seconds = 10 minutes) if at least 5000 Labels need to be processed.
#     """
#     number_of_labels = int(to_barcode) - int(from_barcode)
#     if number_of_labels >= 5000:
#         timeout = int(number_of_labels // 10)  # Assume that mare than ten Labels can be processed per second
#         frappe.enqueue(method=pick_labels, queue='long', timeout=timeout, is_async=True, job_name='pick_labels',
#                     sales_order=sales_order,
#                     from_barcode=from_barcode,
#                     to_barcode=to_barcode)
#         return f"Need to process {number_of_labels} Labels. Please wait a few minutes, go to the Sales Order {sales_order}, open the Delivery Note and print it."
#     else:
#         dn_name = pick_labels(sales_order, from_barcode, to_barcode)
#         return dn_name


@frappe.whitelist()
def pick_labels(sales_order, from_barcode, to_barcode):
    # set flag to prevent running duplicate
    frappe.db.set_value("Sequencing Settings", "Sequencing Settings", "flag_picking_labels", datetime.now().strftime("%Y-%m-%d %H:%M:%S"));
    frappe.db.commit()
    # create sequencing labels
    item = frappe.db.sql("""SELECT `item_code`
        FROM `tabSales Order Item`
        WHERE `parent` = "{sales_order}"
        ORDER BY `idx` ASC
        LIMIT 1;""".format(sales_order=sales_order), as_dict=True)[0]['item_code']
    company = frappe.get_value("Sales Order", sales_order, "company")
    customer = frappe.get_value("Sales Order", sales_order, "customer")
    customer_name = frappe.get_value("Sales Order", sales_order, "customer_name")
    contact = frappe.get_value("Sales Order", sales_order, "contact_person")
    web_order_id = frappe.get_value("Sales Order", sales_order, "web_order_id")
    register_labels = frappe.get_value("Sales Order", sales_order, "register_labels")
    for i in range(int(from_barcode), (int(to_barcode) + 1)):
        # create label
        new_label = frappe.get_doc({
            'doctype': 'Sequencing Label',
            'item': item,
            'label_id': str(i),
            'sales_order': sales_order,
            'customer': customer,
            'customer_name': customer_name,
            'status': "unused",
            'contact': contact,
            'registered_to': contact if register_labels else None
        }).insert()
    frappe.db.commit()
    
    # create delivery note
    dn_content = make_delivery_note(sales_order)

    ## TODO: Consider moving this test before creating the Sequencing Labels
    if len(dn_content.items) == 0:
        frappe.throw(f"Cannot create Delivery Note for {sales_order}. There are no Items left to deliver.")
    
    dn = frappe.get_doc(dn_content)
    dn.naming_series = get_naming_series("Delivery Note", company)
    dn.insert()
    dn.submit()
    frappe.db.commit()
    
    assignment_data = {
            "person_id": contact,
            "sales_order": sales_order,
            "web_order_id": web_order_id,
            "item": item,
            "from_barcode": from_barcode,
            "to_barcode": to_barcode
    }
    write_assignment_file(assignment_data)
    # print address label
    # TODO: see labels.py, take data from delivery note and pass to printer
    
    # reset flag to prevent running duplicate
    frappe.db.set_value("Sequencing Settings", "Sequencing Settings", "flag_picking_labels", None)
    frappe.db.commit()
    
    # return print format
    return dn.name


@frappe.whitelist()
def are_labels_available(item_code, from_barcode, to_barcode):
    conflicts = frappe.db.sql("""
        SELECT `name`
        FROM `tabSequencing Label` 
        WHERE `item` = "{item_code}"
          AND `label_id` BETWEEN "{from_barcode}" AND "{to_barcode}"
          AND LENGTH(`label_id`) = "{length}";
    """.format(item_code=item_code, from_barcode=from_barcode, to_barcode=to_barcode, length=len(from_barcode)), as_dict=True)
    
    if len(conflicts) == 0:
        return 1
    else:
        return 0


@frappe.whitelist()
def picking_ready():
    # Check if the picking labels flag is set (a process is still picking labels)
    flag = frappe.db.get_value("Sequencing Settings", "Sequencing Settings", "flag_picking_labels")
    if not flag:
        return True
    if type(flag) == str:
        flag = datetime.strptime(flag, "%Y-%m-%d %H:%M:%S")
    if (datetime.now() - flag).total_seconds() > 300:
        return True
    return False 
    
